src/app/gibbs/single_thread_sampler.cpp

```diff
@@ -50,109 +50,19 @@ void SingleThreadSampler::sample_sgd_single_variable(long vid) {
-
-    *this->p_rand_obj_buf = erand48(this->p_rand_seed);
-    if ((*this->p_rand_obj_buf) *
-            (1.0 + exp(potential_neg_freeevid - potential_pos_freeevid)) <
-        1.0) {
-      p_fg->template update<true>(variable, 1.0);
-    } else {
-      p_fg->template update<true>(variable, 0.0);
-    }
-
-    this->p_fg->update_weight(variable);
-
-  } else if (variable.domain_type == DTYPE_MULTINOMIAL) {  // multinomial
-
-    // varlen_potential_buffer contains potential for each proposals
-    while (variable.upper_bound >= varlen_potential_buffer.size()) {
-      varlen_potential_buffer.push_back(0.0);
-    }
-
-    if (variable.is_evid == false) {
-      sum = -100000.0;
-      acc = 0.0;
-      multi_proposal = -1;
-
-      // calculate potential for each proposal
-      for (int propose = variable.lower_bound; propose <= variable.upper_bound;
-           propose++) {
-        varlen_potential_buffer[propose] =
-            p_fg->template potential<false>(variable, propose);
-        sum = logadd(sum, varlen_potential_buffer[propose]);
-      }
-
-      // flip a coin
-      *this->p_rand_obj_buf = erand48(this->p_rand_seed);
-      for (int propose = variable.lower_bound; propose <= variable.upper_bound;
-           propose++) {
-        acc += exp(varlen_potential_buffer[propose] - sum);
-        if (*this->p_rand_obj_buf <= acc) {
-          multi_proposal = propose;
-          break;
-        }
-      }
-      assert(multi_proposal != -1);
-      p_fg->update_evid(variable, multi_proposal);
-    }
-
-    sum = -100000.0;
-    acc = 0.0;
-    multi_proposal = -1;
-    for (int propose = variable.lower_bound; propose <= variable.upper_bound;
-         propose++) {
-      varlen_potential_buffer[propose] =
-          p_fg->template potential<true>(variable, propose);
-      sum = logadd(sum, varlen_potential_buffer[propose]);
-    }
-
-    *this->p_rand_obj_buf = erand48(this->p_rand_seed);
-    for (int propose = variable.lower_bound; propose <= variable.upper_bound;
-         propose++) {
-      acc += exp(varlen_potential_buffer[propose] - sum);
-      if (*this->p_rand_obj_buf <= acc) {
-        multi_proposal = propose;
-        break;
-      }
-    }
-    assert(multi_proposal != -1);
-    p_fg->template update<true>(variable, multi_proposal);
+  // sample the variable with evidence unchanged
+  if (variable.is_evid == false) {
+    proposal = draw_sample(variable, false);
+    p_fg->update_evid(variable, (double)proposal);
+  }
 
-    this->p_fg->update_weight(variable);
+  // sample the variable regardless of whether it's evidence
+  proposal = draw_sample(variable, true);
+  p_fg->template update<true>(variable, (double)proposal);
 
-  } else {
-    // std::cout << "~~~~~~~~~" << std::endl;
-    std::cerr
-        << "[ERROR] Only Boolean and Multinomial variables are supported now!"
-        << std::endl;
-    assert(false);
-    return;
-  }  // end if for variable types
```

Comment:
Was this else part dropped intentionally?  Can't find it in the new `draw_sample()`.